### ISSUE TITLE
Investigate potential crash from the community chat

### ISSUE THREAD

**Issue by Gerold103** [tarantool/tarantool#6837]
Here is the original message:
```
Если вызвать fiber_start()
И сразу за ним fiber_cancel()
То частенько бывает такой вот креш:

Tarantool 2.8.3-0-g01023dbc2

2022-02-04 09:32:41.326 [1] main/119/hello I> [0x7f91ccde3b80:0x7f91cc801210] fiber_fun
 Segmentation fault
   code: SEGV_MAPERR
   addr: (nil)
   context: 0x7f91b3b7fa00
   siginfo: 0x7f91b3b7fb30
   rax      0x0                0
   rbx      0x7f91cc801210     140264177930768
   rcx      0x5611078008e4     94631140264164
   rdx      0x5611099ffbc0     94631175912384
   rsi      0x7f91cc8012f8     140264177931000
   rdi      0x7f91cc8012f8     140264177931000
   rsp      0x7f91b3b7ff98     140263762165656
   rbp      0x7f91b3b7ffa0     140263762165664
   r8       0x5611099fff38     94631175913272
   r9       0x67               103
   r10      0x8000000          134217728
   r11      0xf60e             62990
   r12      0x7f91cc8012e0     140264177930976
   r13      0x0                0
   r14      0x0                0
   r15      0x0                0
   rip      0x0                0
   eflags   0x10206            66054
   cs       0x33               51
   gs       0x0                0
   fs       0x0                0
   cr2      0x0                0
   err      0x14               20
   oldmask  0x0                0
   trapno   0xe                14
 Current time: 1643967161
 Please file a bug at http://github.com/tarantool/tarantool/issues
 Attempting backtrace... Note: since the server has already crashed, 
 this may fail as well
 #0  0x5611077f9c24 in crash_signal_cb+94
 #1  0x7f91cd158730 in funlockfile+50
 #2  (nil) in +50
 #3  0x561107800970 in fiber_loop+30
 #4  0x561107ce0c5f in coro_init+3f
```
Need to check if this is a bug on our side.

The code:
```C++
//
// Created by Andrey Syvrachev on 20.10.2021.
//
#include <module.h>
#include <pthread.h>


static int fiber_fun_test(va_list) {
    say_info("[%p:%p] fiber_fun_test", pthread_self(), fiber_self());
    return 0;
}

LUA_API "C" int luaopen_fiber_crash_demo(lua_State *L) {
    say_info("[%p:%p] fiber_new", pthread_self(), fiber_self());
    auto fiber = fiber_new("hello", fiber_fun_test);
    say_info("[%p:%p] fiber_start", pthread_self(), fiber_self());
    fiber_start(fiber);
    say_info("[%p:%p] fiber_cancel", pthread_self(), fiber_self());
    fiber_cancel(fiber); // just comment fiber_cancel to remove crash
    say_info("[%p:%p] fiber_cancel .. fin", pthread_self(), fiber_self());
    return 1;
}
```
from here https://github.com/allright/tarantool_fiber_crash_demo/blob/master/cpp/fiber_crash_demo.cpp.

**Comment by Gerold103** [comment]
The reason is that `fiber_start()` finishes the fiber right away because the function in the example (`fiber_fun_test()`) doesn't yield. After `fiber_start()` the fiber is already recycled. Calling `fiber_cancel()` on a recycled fiber is almost the same as use-after-free. For instance, its C function pointer is already nullified. The crash in the ticket is because `fiber_start()` tries to call a zero pointer.

Nonetheless there is a solution which would make it "work". Meaning that it won't crash in Release, it will be just NOP. But it will assert in Debug. I have it working locally.

NOP means that it depends on your luck. If between return from `fiber_start()` and `fiber_cancel()` you didn't have any yields - it will be NOP indeed. Otherwise if the fiber is reused, you will cancel some other random fiber.

I need a decision what to do. Either close this as not our bug, or made the aforementioned "fix". Or propose something else. I ping @sergos since he is probably more aware of such low level API than the product managers.

**Comment by sergos** [comment]
First of all in your solution the assertion becomes incorrect, obviously. I didn't see the , can't comment exactly why.

In my opinion, the `fiber_loop()` behaves incorrectly in part of making decision on recycling: it issues madvise of `MADV_DONTNEED` on  the stack memory, that is equivalent of zero-initialization - only by chance not immediate - while keep execution in the very same context AFAIU. 

As for call to the `fiber_cancel(fiber);` in the test - it is clearly a misuse for the case `fiber_new()` runs a function that ends without a yield. The same can be achieved, if you make two subsequent calls to the `fiber_cancel()` using the same fiber argument - the first one will successfully cleanup - and potentially reuse - the fiber, the second one will be called on an empty - or as you mentioned on an arbitrary - fiber context. The behavior of the `fiber_new()` is described in the code comment and on the [site](https://www.tarantool.io/en/doc/latest/dev_guide/reference_capi/fiber/#c.fiber_new). Unfortunately, we can't help with such a behavior, unless we introduce an identification for fibers (say, monotonic grow id) and separate the user-visible `fiber` structure from the scheduler one.

To sum it up: `fiber_loop()` has to postpone cleanup of the fiber structure until it is switched to the scheduler; the `fiber_cancel()` should assert a non-zero `fid` only after it checks for `fiber_is_dead()`.

**Comment by sergos** [comment]
I would like to come up with a complete resolution rather than partial fixup, which is the 2nd bullet of your plan looks like. The risky part of it looks very bad, imho.

**Comment by Gerold103** [comment]
It can not be reliably fixed for all use cases. Firstly, `fiber_cancel()` can't fail, it is `void`. Secondly, even if we do add a panic or a log in `fiber_cancel()` for a fiber which was already recycled, it won't work in case there was a yield between fiber's death and `fiber_cancel()` - the fiber could already be reused, so `fiber_cancel()` will get a perfectly valid fiber pointer with all fields set correctly.

There simply is no way to tell if you called `fiber_cancel()` on the fiber you wanted. It is not possible. Even if we would make users create their own fiber wrappers which are never reused and are properly freed, then we end up with a potential use-after-free issue. It can't be fixed in C.

There are only 2 possible solutions: either apply my patch as is, or add a `panic()` in `fiber_cancel()` for the fibers which are recycled. In the latter case it should catch the issue almost always soon or late on the user's side with a proper message instead of a crash.
